File: <100/064/064.cpp
#include<iostream>
#include<set>
#include<vector>
#include<tuple>
#include<algorithm>
#include<numeric>
using namespace std;
using TP = tuple<int, int, int>;


const int max_N = 1e4, max_binary = 101;
int answer;


// smaller*smaller <= target < bigger*bigger
int findSqrt(int smaller, int target, int bigger){
  if(smaller + 1 == bigger) return smaller;
  int mid = (smaller + bigger) / 2;
  int m2 = mid * mid;
  if(m2 <= target) return findSqrt(mid, target, bigger);
  else             return findSqrt(smaller, target, mid);
}


TP findFirst(int n){
  int s = findSqrt(0, n, max_binary);
  TP t = make_tuple(n, s, n-s*s);
  return t;
}


int findPeriod_(vector<TP> &V, TP t){
  auto itr = find(V.begin(), V.end(), t);
  int p = distance(itr, V.end());
  return p;
}


int findPeriod(TP tk, vector<TP> &V, set<TP> &S){
  // check supplemental.md to see the meanings of variables and the strategy.
  auto [nk, bk, dk] = tk;
  int sk = findSqrt((int)0, nk, max_binary) + bk;
  int ak = sk / dk;
  int ck = dk * ak - bk;

  int dk1 = nk - ck*ck;
  int hcf = gcd(dk, dk1);
  dk /= hcf,  dk1 /= hcf;

  int nk1 = dk * dk * nk;
  int bk1 = dk * ck;
  TP tk1 = make_tuple(nk1, bk1, dk1);
  if(S.count(tk1) == 1) return findPeriod_(V, tk1);
  
  S.insert(tk1);
  V.push_back(tk1);
  return findPeriod(tk1, V, S);
}


int main(){
  for(int i = 2; i <= max_N; ++i){
    TP t1 = findFirst(i);
    auto [_1, _2, d] = t1;
    // d == 0  <==>  i is squared number such as 4 or 9
    if(d == 0) continue;

    vector<TP> V;
    set<TP> S;
    V.push_back(t1);
    S.insert(t1);
    int p = findPeriod(t1, V, S);
    answer += p&1;
  }

  cout << answer << endl;
  return 0;
}

File: <100/064/064.jl
function sqrtInt(n :: Int, smaller :: Int, bigger :: Int) :: Int
  if bigger - smaller == 1
    return smaller
  end
  mid = (bigger + smaller) ÷ 2
  s = mid^2
  if s <= n
    return sqrtInt(n, mid, bigger)
  else
    return sqrtInt(n, smaller, mid)
  end
end


function findFirst(n :: Int) :: Tuple{Int, Int, Int}
  s = sqrtInt(n, 0, 10^9)
  return (n, s, n-s^2)
end


function find_period_(v :: Vector{Tuple{Int, Int, Int}}, tk :: Tuple{Int, Int, Int}) :: Int
  l = length(v)
  for i ∈ 1:l
    if v[i] == tk
      return l - i + 1
    end
  end
  return -1
end


function find_period(tk :: Tuple{Int, Int, Int}, v :: Vector{Tuple{Int, Int, Int}}, s :: Set{Tuple{Int, Int, Int}}) :: Int
  (nk, bk, dk) = tk
  sk = sqrtInt(nk, 0, 10^9) + bk
  ak = sk ÷ dk
  ck = dk * ak - bk

  dk1 = nk - ck^2
  hcf = gcd(dk, dk1)
  dk  = dk ÷ hcf
  dk1 = dk1 ÷ hcf

  nk1 = dk^2 * nk
  bk1 = dk * ck
  tk1 = (nk1, bk1, dk1)

  if tk1 ∈ s
    return find_period_(v, tk1)
  end
  push!(s, tk1)
  push!(v, tk1)
  return find_period(tk1, v, s)
end


function main() :: Nothing
  upper_limit = 10^4
  answer = 0
  for i ∈ 2:upper_limit
    t1 = findFirst(i);
    d = t1[3]
    if d == 0
      continue
    end

    v :: Vector{Tuple{Int, Int, Int}} = [t1]
    s :: Set{Tuple{Int, Int, Int}}    = Set([t1])
    p = find_period(t1, v, s)
    answer += p&1
  end

  answer |> println
end


@time main()
